### PR TITLE
New signal API: trio.open_signal_receiver

### PR DESCRIPTION
**Pull request by njsmith**
Fixes gh-354

Other changes:

- deprecate trio.catch_signal
- fix a few small edge-cases I noticed along the way

### CODE REVIEW

File: trio/tests/test_signals.py [hunk 1]
```diff
@@ -35,23 +34,98 @@
...
+    orig = signal.getsignal(signal.SIGILL)
+    with open_signal_receiver(signal.SIGILL) as receiver:
+        # Raise it a few times, to exercise signal coalescing, both at the
+        # call_soon level and at the SignalQueue level
+        signal_raise(signal.SIGILL)
+        signal_raise(signal.SIGILL)
+        await _core.wait_all_tasks_blocked()
+        signal_raise(signal.SIGILL)
+        await _core.wait_all_tasks_blocked()
+        async for signum in receiver:  # pragma: no branch
```
Comment (tgoodlet) [inline trio/tests/test_signals.py]: Is it worth ensuring that `receiver` only *once* delivers a signal raised multiple times prior to iteration?
I guess I'm just thinking down the road if the implementation is changed (internal `set`-like data structure or whatever) and this behaviour should remain.

File: trio/tests/test_signals.py [hunk 2]
```diff
@@ -60,50 +134,44 @@ def test_catch_signals_wrong_thread():
     def direct_handler(signo, frame):
         delivered_directly.add(signo)
 
-    async def wait_call_soon_idempotent_queue_barrier():
-        ev = Event()
-        token = _core.current_trio_token()
-        token.run_sync_soon(ev.set, idempotent=True)
-        await ev.wait()
-
     print(1)
```
Comment (tgoodlet) [inline trio/tests/test_signals.py]: Maybe print something more descriptive?

Comment (tgoodlet) [reply trio/tests/test_signals.py]: Hmm, actually, have you considered just adding some branch logic for each 
`await wait_run_sync_soon_idempotent_queue_barrier()` location and then just use `pytest.mark.parametrize()` over signals and that?

Might reduce some code unless the duplication is intentional.

File: trio/_signals.py [hunk 3]
```diff
@@ -1,11 +1,15 @@
 import signal
 from contextlib import contextmanager
+from collections import OrderedDict
```
[the remark this answers is not in the document]
Comment (njsmith) [reply trio/_signals.py]: PyPy has actually had ordered `dict`s by default for a while (CPython copied their implementation). But annoyingly, we also need `popitem(last=False)`, which even in 3.7 is still an `OrderedDict`-only method.

File: trio/tests/test_signals.py [hunk 4]
```diff
@@ -1,17 +1,16 @@
...
+import trio
 from .. import _core
 from .._util import signal_raise
-from .._signals import catch_signals, _signal_handler
-from .._sync import Event
+from .._signals import catch_signals, open_signal_receiver, _signal_handler
 
 
-async def test_catch_signals():
+# Delete when catch_signals is removed
```
[the remark this answers is not in the document]
Comment (njsmith) [reply trio/tests/test_signals.py]: Eh, when we remove `catch_signals` the test will start failing, so we won't miss it :-)

File: trio/_signals.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (tgoodlet) [reply trio/_signals.py]: >which even in 3.7 is still an OrderedDict-only method.

Ahh what!?

File: trio/tests/test_signals.py (hunk 1 above)
Comment (njsmith) [reply trio/tests/test_signals.py]: We are implicitly checking that signals are coalesced, because if they weren't then when we exited the `with` block they'd be redelivered and cause the test process to dump core :-).

It would make sense to test this a little more explicitly... I'm going to add a little private helper to do that.

File: trio/tests/test_signals.py (hunk 2 above)
Comment (njsmith) [reply trio/tests/test_signals.py]: This is a pretty complicated test that I didn't really touch here, so honestly I'm not sure what it does right now :-). I doubt it's worth messing with – if it breaks we can easily add better print statements or whatever then.

**Comment by njsmith** [comment]
Added a more explicit test of signal coalescing. All comments addressed.

File: trio/tests/test_signals.py (hunk 1 above)
Comment (tgoodlet) [reply trio/tests/test_signals.py]: Ahh right of course. Cool, I like explicit :+1:

**Comment by pquentin** [comment]
Merging since @tgoodlet approved this
